test-d/replace-optional-deep.ts

```diff
@@ -0,0 +1,118 @@
+import {expectType} from 'tsd';
+import type {ReplaceOptionalDeep} from '../index';
+
+type In1 = {a: string; b?: boolean};
+type Out1 = {a: string; b: undefined | boolean};
+
+type In2 = {a?: string; b?: boolean};
+type Out2 = {a: null | string; b: null | boolean};
+
+type In3 = {a: string; b: boolean};
+type Out3 = {a: string; b: boolean};
+
+type In4 = {a: undefined};
+type Out4 = In4;
+
+type In5 = {
+	a: 1;
+	b?: 2;
+	c?: undefined | 3;
+	d?: null | 4 ;
+	e?: undefined | null | 5;
+	f: undefined | 6;
+	g?: {};
+	h: {
+		i: 7;
+		j?: 7;
+		k?: {
+			l?: 8;
+			m: 9;
+			n?: undefined | 10;
+			o?: null | 11;
+			p?: undefined | null | 12;
+			q: undefined | 13;
+		};
+	};
+};
+
+type Out5 = {
+	a: 1;
+	b: undefined | 2;
+	c: undefined | 3;
+	d: undefined | null | 4;
+	e: undefined | null | 5;
+	f: undefined | 6;
+	g: undefined | {};
+	h: {
+		i: 7;
+		j: undefined | 7;
+		k:
+		| undefined
+		| {
+			l: undefined | 8;
+			m: 9;
+			n: undefined | 10;
+			o: undefined | null | 11;
+			p: undefined | null | 12;
+			q: undefined | 13;
+		};
+	};
+};
+
+type In6 = {
+	a?:
+	| 1
+	| {
+		b: 2;
+		c?: 3;
+		d:
+		| undefined
+		| readonly undefined[]
+		| {
+			e?:
+			| {a?: 1}
+			| {a?: 2};
+		};
+	};
+};
+
+type Out6 = {
+	a:
+	| undefined
+	| 1
+	| {
+		b: 2;
+		c:
+		| undefined
+		| 3;
+		d:
+		| undefined
+		| readonly undefined[]
+		| {
+			e:
+			| undefined
+			| {a: undefined | 1}
+			| {a: undefined | 2};
+		};
+	};
+};
+
+type In7 = {a?: 1};
+
+type Out7 = {a: 0 | 1};
+
+declare const test1: ReplaceOptionalDeep<In1>;
+declare const test2: ReplaceOptionalDeep<In2, null>;
+declare const test3: ReplaceOptionalDeep<In3>;
+declare const test4: ReplaceOptionalDeep<In4>;
+declare const test5: ReplaceOptionalDeep<In5>;
+declare const test6: ReplaceOptionalDeep<In6>;
+declare const test7: ReplaceOptionalDeep<In7, 0>;
+
+expectType<Out1>(test1);
+expectType<Out2>(test2);
+expectType<Out3>(test3);
+expectType<Out4>(test4);
+expectType<Out5>(test5);
+expectType<Out6>(test6);
+expectType<Out7>(test7);
```

Comment:
Peanut gallery observation: visually verifying this code is going to require scrolling up and down a lot, particularly if this grows more complex in the future.  I would recommend moving the declares and expectTypes closer to the types you're actually testing.  But if the project owners disagree, I would go with their recommendation.

Not-so-peanut-gallery observation: what happens if someone puts the find type inside the replacement type?  Let's come up with some evil replacement testcases too, including:

- complex types for the replace value
- `never`
- `unknown`
- `any`
- a value already in the type

Reply:
@ajvincent thank you for your review.

I appreciate you putting thought into how recursion could misbehave if the replacement itself contained part(s) that matched the needle. The nice thing about implementing find/replace this way (handling a positive match as a base case) is that you avoid that complexity altogether.

I added extra tests for `Patch`, which I'll be requesting your review on in the next day or 2. But wanted to answer your concern while it was fresh on my mind -- since a match is treated as a base case, and we don't continue recursing, users can use any type they like: `any`, `never`, a copy of the entire tree -- because we consider that to be the base case, we will never traverse down that subtree.